### PR TITLE
feat(imgproc): native CUDA downscale kernels — __ldg, texture objects, fused normalise

### PR DESCRIPTION
**Pull request by Incharajayaram**
## 📝 Description

**Fixes/Relates to:** (GSoC 2026 — GPU-accelerated image processing deliverable)

Adds native CUDA downscale kernels for 3-channel f32 images via NVRTC, reaching
parity with PyTorch `F.interpolate` bilinear and eliminating the overhead of a
separate normalise pass for vision pipelines.

---

## 🛠️ Changes Made

- [x] **`crates/kornia-imgproc/src/gpu/resize_cuda.rs`** — new module:
  - `launch_resize_bilinear_downscale_cuda`: bilinear downscale via NVRTC + `__ldg` read-only cache, 32×8 thread block, `CU_FUNC_CACHE_PREFER_L1`; ~70 GB/s on GTX 1650 (matches PyTorch)
  - `launch_resize_nearest_downscale_cuda`: nearest-neighbor downscale; ~116 GB/s (~91% DRAM peak)
  - `launch_resize_bilinear_normalize_cuda`: fused bilinear downscale + per-channel `(pixel − mean) / std` in a single kernel pass, eliminating a separate normalise DRAM round-trip (~1.7× speedup for resize+normalise pipelines)
  - `CudaResizeError` error type
  - Module doc explains 32×8 vs 16×16 block choice, why shared-memory tiling was ruled out, roofline analysis, and throughput table

- [x] **`crates/kornia-imgproc/src/gpu/resize_cuda_tex.rs`** — new module:
  - `launch_resize_bilinear_tex_cuda`: hardware bilinear via `cudaTextureObject_t` + `CU_RESOURCE_TYPE_PITCH2D` + `CU_TR_FILTER_MODE_LINEAR`; eliminates software weight computation
  - `CudaRgbTexture`: de-interleaves RGB → planar planes and binds each as a pitch2D texture
  - `create_pitch2d_texture`, `query_pitch_alignment` helpers; `CudaTexResizeError` error type
  - Useful for multi-resize pipelines where setup cost is amortised

- [x] **`crates/kornia-tensor/src/cuda.rs`** — `prefer_l1_cache()` added to `CudaKernel`; sets `CU_FUNC_CACHE_PREFER_L1` to enlarge L1 from 32 KB → 64 KB on Turing

- [x] **`crates/kornia-imgproc/src/gpu/mod.rs`** — `resize_cuda` and `resize_cuda_tex` modules registered under `#[cfg(feature = "gpu-cuda")]`

- [x] **`crates/kornia-imgproc/examples/bench_gpu_resize.rs`** — three new benchmark sections:
  - `run_gpu_cuda`: nearest + bilinear kernels across 3 downscale resolutions with GB/s column
  - `run_gpu_cuda_tex`: texture-object kernel with setup vs kernel-only timing separated
  - `run_gpu_cuda_fused_normalize`: fused kernel vs resize-only side-by-side

---

## 🧪 How Was This Tested?

- [x] **Manual Verification:** all three bench sections run on GTX 1650 (CUDA 12.4, sm_75); output verified against PyTorch `F.interpolate(mode='bilinear', align_corners=False)` — pixel values match to f32 rounding tolerance
- [x] **Performance/Edge Cases:**
  - Boundary pixels (edges, corners) clamped correctly via `fmaxf/fminf` and `min(x+1, w-1)`
  - `std == 0` guarded in `launch_resize_bilinear_normalize_cuda`
  - Pitch alignment queried from device via `CU_DEVICE_ATTRIBUTE_TEXTURE_PITCH_ALIGNMENT`; error returned if source width produces misaligned pitch
  - Kernels gated behind `#[cfg(feature = "gpu-cuda")]` — CI on CPU-only and aarch64 runners unaffected

**Measured throughput (GTX 1650, 2× downscale, 200 iterations):**

| Kernel                   | 1080p→540p | GB/s  |
|--------------------------|-----------|-------|
| Nearest (`__ldg`)        | 0.107 ms  | ~116  |
| Bilinear (`__ldg`)       | 0.178 ms  | ~70   |
| Bilinear + normalise     | 0.178 ms  | ~70   |
| Bilinear (texture HW)    | 0.165 ms  | ~76   |
| CPU bilinear (kornia)    | 4.858 ms  | ~2.6  |

The bilinear kernel reaches the DRAM read+write turnaround ceiling for GDDR5 mixed-traffic (~55–60% of 128 GB/s bidirectional peak); no further software optimisation is possible for this data layout and dtype.

---

## 🕵️ AI Usage Disclosure

- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.

---

## 🚦 Checklist

- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.

---

## 💭 Additional Context

- The `gpu-cuda` feature depends on `cudarc` (already a workspace dep from PR #950). No new dependencies added.
- Texture-object path (`resize_cuda_tex`) is a separate module intentionally — it requires a de-interleave preprocessing step and different setup semantics vs the inline `__ldg` kernels.
- Shared-memory tiling was evaluated and found to be slower for 2× downscale (L2 already handles bilinear reuse; smem adds overhead without reducing DRAM reads). The module doc explains the roofline analysis in full.

### CODE REVIEW

File: crates/kornia-imgproc/src/gpu/resize_cuda.rs [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+    dst[out]     = __ldg(&src[src_base]);
+    dst[out + 1] = __ldg(&src[src_base + 1]);
+    dst[out + 2] = __ldg(&src[src_base + 2]);
+}
+"#;
+
+// ── CUDA C source: fused bilinear downscale + per-channel normalise ───────────
+
+static BILINEAR_NORMALIZE_SRC: &str = r#"
+extern "C" __global__ void resize_bilinear_normalize_3c(
```
Comment (edgarriba) [inline crates/kornia-imgproc/src/gpu/resize_cuda.rs:146]: would be great a variant also that does in the same fused op HWC -> CHW

**Review by edgarriba — APPROVED** [review]
LGTM. Add also the benchmarks results somewhere persistant

**Comment by edgarriba** [comment]
@Incharajayaram resolve conficts

**Comment by edgarriba** [comment]
Superseded by #958, which already merged this branch's content (plus warp-affine kernels) into main on 2026-06-30. Compare shows 0 file diff against main (`gh api repos/kornia/kornia-rs/compare/main...Incharajayaram:kornia-rs:gpu/resize-downscale` → files: 0) — nothing left to merge here.
